server/src/main/java/org/opensearch/wlm/WorkloadManagementTransportInterceptor.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.wlm;
+
+import org.opensearch.tasks.Task;
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.transport.TransportChannel;
+import org.opensearch.transport.TransportInterceptor;
+import org.opensearch.transport.TransportRequest;
+import org.opensearch.transport.TransportRequestHandler;
+
+/**
+ * This class is used to intercept search traffic requests and populate the queryGroupId header in task headers
+ */
+public class WorkloadManagementTransportInterceptor implements TransportInterceptor {
+    private final ThreadPool threadPool;
+
+    public WorkloadManagementTransportInterceptor(ThreadPool threadPool) {
+        this.threadPool = threadPool;
+    }
+
+    @Override
+    public <T extends TransportRequest> TransportRequestHandler<T> interceptHandler(
+        String action,
+        String executor,
+        boolean forceExecution,
+        TransportRequestHandler<T> actualHandler
+    ) {
+        return new WorkloadManagementTransportHandler<T>(threadPool, actualHandler);
+    }
+
+    /**
+     * This class is mainly used to populate the queryGroupId header
+     * @param <T> T is Search related request
+     */
+    public static class WorkloadManagementTransportHandler<T extends TransportRequest> implements TransportRequestHandler<T> {
```

Comment:
We can rename this to RequestHandler? WorkloadManagementTransport is clear from the outerclass.